File: spring-skeleton/src/main/java/com/example/springskeleton/adapter/member/api/MemberApiController.java
package com.example.springskeleton.adapter.member.api;

import com.example.springskeleton.application.MemberService;
import com.example.springskeleton.domain.member.Member;
import com.example.springskeleton.dto.member.MemberDto;
import com.example.springskeleton.dto.member.MemberSearchCondition;
import com.example.springskeleton.exception.MemberException;
import lombok.RequiredArgsConstructor;
import org.springframework.web.bind.annotation.GetMapping;
import org.springframework.web.bind.annotation.PostMapping;
import org.springframework.web.bind.annotation.RequestParam;
import org.springframework.web.bind.annotation.RestController;

import java.util.List;

@RestController
@RequiredArgsConstructor
public class MemberApiController {

    private final MemberService memberService;

    @GetMapping("api/v1/members")
    public List<Member> getMembersWithCondition(@RequestParam(value = "offset", defaultValue = "0") int limit,
                                   @RequestParam(value = "limit", defaultValue = "100") int offset) {
        MemberSearchCondition condition = MemberSearchCondition.builder()
                .limit(limit)
                .offset(offset)
                .build();
        return memberService.findMembersWith(condition);
    }

    @GetMapping("api/v1/member")
    public Member getMember(@RequestParam("memberId") Long memberId) throws MemberException {
        return memberService.getMember(memberId);
    }

    @PostMapping("api/v1/member")
    public Long joinMember(MemberDto memberDto) throws MemberException {
        return memberService.join(memberDto);
    }


}

File: spring-kotlin-user/src/main/kotlin/com/example/kotlin/user/presentation/MemberApiController.kt
package com.example.kotlin.user.presentation

import com.example.kotlin.user.business.MemberService
import com.example.kotlin.user.business.domain.Member
import com.example.kotlin.user.common.authority.TokenInfo
import com.example.kotlin.user.common.dto.BaseResponse
import com.example.kotlin.user.common.dto.CustomUser
import com.example.kotlin.user.common.util.toDtoResponse
import com.example.kotlin.user.dto.LoginDto
import com.example.kotlin.user.dto.MemberDtoRequest
import com.example.kotlin.user.dto.MemberDtoResponse
import jakarta.validation.Valid
import org.springframework.security.core.context.SecurityContextHolder
import org.springframework.web.bind.annotation.*
import org.springframework.security.access.AccessDeniedException

@RequestMapping("/api/members")
@RestController
class MemberApiController(
    private val memberService: MemberService
) {
    /**
     * 회원 가입
     */
    @PostMapping("/v1/sign-up")
    fun signUpV1(@RequestBody @Valid memberDtoRequest: MemberDtoRequest): MemberDtoResponse {
        val member: Member = memberService.signUp(memberDtoRequest = memberDtoRequest)
        return member.toDtoResponse()
    }

    @PostMapping("/v2/sign-up/")
    fun signUpV2(@RequestBody @Valid memberDtoRequest: MemberDtoRequest): BaseResponse<MemberDtoResponse> {
        val member: Member = memberService.signUp(memberDtoRequest = memberDtoRequest)
        return BaseResponse(
            message = "회원 가입 성공",
            data = member.toDtoResponse()
        )
    }

    /**
     * 로그인
     */
    @PostMapping("/login")
    fun login(@RequestBody @Valid loginDto: LoginDto): BaseResponse<TokenInfo> {
        val tokenInfo = memberService.login(loginDto)
        return BaseResponse(data = tokenInfo)
    }

    /**
     * 내 정보 조회하기
     * - 문제점: 다른 유저의 정보까지 조회가 가능 함
     */
    @GetMapping("/info1/{id}")
    fun searchMyInfo(@PathVariable id: Long): BaseResponse<MemberDtoResponse> {
        val member = memberService.searchMyInfo(id)
        return BaseResponse(data = member.toDtoResponse())
    }

    /**
     * 내 정보 조회하기 v2
     * - 문제점: 다른 유저의 정보까지 조회가 가능 함
     */
    @GetMapping("/info2/{id}")
    fun searchMyInfoV2(@PathVariable id: Long): BaseResponse<MemberDtoResponse> {
        val userId = (SecurityContextHolder
            .getContext()
            .authentication
            .principal as CustomUser)
            .userId

        if (userId != id) {
            throw AccessDeniedException("해당 정보에 접근할 수 없습니다.")
        }
        val member = memberService.searchMyInfo(id)
        return BaseResponse(data = member.toDtoResponse())
    }

    /**
     * 내 정보 조회하기
     * - 로그인한 사람의 정보만 조회할 수 있도록 수정
     */
    @GetMapping("/info3")
    fun searchMyInfo(): BaseResponse<MemberDtoResponse> {
        val userId = (SecurityContextHolder
            .getContext()
            .authentication
            .principal as CustomUser)
            .userId
        val member = memberService.searchMyInfo(userId)

        return BaseResponse(data = member.toDtoResponse())
    }

    /**
     * 회원 정보 수정하기
     */
    @PutMapping("/info")
    fun modifyMyInfo(@RequestBody @Valid memberDtoRequest: MemberDtoRequest): BaseResponse<MemberDtoResponse> {
        val userId = (SecurityContextHolder
            .getContext()
            .authentication
            .principal as CustomUser)
            .userId
        memberDtoRequest.id = userId
        val member = memberService.modifyMyInfo(memberDtoRequest)
        return BaseResponse(data = member.toDtoResponse())
    }
}

